### DISCUSSION TITLE
Can't open include file "NPR_Pipeline.glsl"

### DISCUSSION THREAD

**Discussion by Floreum** [bnpr/Malt#533] [Q&A] [answered]
Hello, I was following the nodes to code documentation and I cannot get the example code to work. I tried looking around for an answer here but I can't seem to find a solution from the various threads asking about this. I'm following the example code given here
[](https://malt3d.com/From-Nodes-To-Code/From-Nodes-To-Code/) so I'm not sure if this is outdated or maybe it's broken in Blender 4.0?

Here's the full error:
```
J:/3D Modeling/2023/Malt_Test/test.mesh.glsl:1: error: Can't open include file "NPR_Pipeline.glsl"

    #include "NPR_Pipeline.glsl"

    from C:/Users/Joel/AppData/Local/Temp/tmpss_z8cf4: 1:    #include "test.mesh.glsl"

1 error in preprocessor.
```

**Comment by pragma37** [comment, marked as answer]
That tutorial is quite old.
It's from the time when Malt didn't have nodes and I still thought I could convince artists to use code-based shaders.😅
Sorry for the confusion.

Now the intended workflow is to code your own custom nodes if you want to use GLSL.
See https://malt3d.com/documentation/plugins/

You can drop functions at [`PluginExample.glsl`](https://github.com/bnpr/Malt/blob/Development/plugins/PluginExample/Shaders/PluginExample.glsl) and they will be available automatically as a node.

So the first examples would become something like this:
```glsl
#include "Common.glsl"

void starting_point()
{

}

void flat_color(vec3 color, out vec3 result)
{
    result = color;
}

void textures(int uv_channel, sampler2D color_texture, out vec4 result)
{
    vec2 texture_coordinates = UV[uv_channel];
    
    vec4 sampled_color = texture(color_texture, texture_coordinates);
    
    result = sampled_color;
}
```

The other two would be much more complex now, since there's not an API aimed at artists anymore.
The lighting one would be something likes this:
```glsl
#include "NPR_Lighting.glsl"

void lighting(vec3 color, out vec3 result)
{
    result = vec3(0);
    for (int i = 0; i < LIGHTS.lights_count; i++)
    {
        Light L = LIGHTS.lights[i];
        LitSurface LS = npr_lit_surface(POSITION, NORMAL, 0, L, i, true, true);
        result += max(LS.NoL, 0.0);
    }
    
    result *= color;
}
```
Which requires way more context to understand.

**Reply by bertodelrio256** [reply]
its funny because I installed MALT because i can write glsl code. (i hate node systems). still not clear where to define funcs like npr_lit_surface() yourself. hopefully i dont have to write a pipeline.

**Reply by bertodelrio256** [reply]
ohhh ok, so I would have to write a plugin to do custom diffuse and specular?
